ksqldb-functional-tests/src/test/resources/query-validation-tests/elements.json

```
@@ -859,6 +859,52 @@
           {"name": "OUTPUT", "type": "stream", "schema": "ID INT KEY, ID_COPY INT"}
         ]
       }
+    },
+    {
+      "name": "table without primary key fails",
+      "statements": [
+        "CREATE TABLE INPUT (ID INT, F0 INT) WITH (kafka_topic='input', value_format='JSON');"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "Tables require a PRIMARY KEY. Please define the PRIMARY KEY."
+      }
+    },
+    {
+      "name": "table without primary key fails - schema inference",
+      "statements": [
+        "CREATE TABLE INPUT WITH (kafka_topic='input', value_format='Avro');"
+      ],
+      "topics": [
+        {
+          "name": "input",
+          "schema": {"name": "blah", "type": "record", "fields": [{"name": "c1", "type": "int"}]},
+          "format": "AVRO"
+        }
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlException",
+        "message": "Tables require a PRIMARY KEY. Please define the PRIMARY KEY.\nUse a partial schema to define the primary key and still load the value columns from the Schema Registry, for example:\n\tCREATE TABLE INPUT (ID INT PRIMARY KEY) WITH (...);"
+      }
+    },
+    {
+      "name": "stream without key column",
+      "statements": [
+        "CREATE STREAM INPUT (ID INT, F0 INT) WITH (kafka_topic='input', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT;"
+      ],
+      "inputs": [
+        {"topic": "input", "key": null, "value": {"ID": 1, "F0": 2}}
```

```suggestion
        {"topic": "input", "key": null, "value": {"ID": 1, "F0": 2}},
        {"topic": "input", "key": "foo", "value": {"ID": 1, "F0": 2}}
```
let's also test what happens when there is a key in the message but one isn't declared

---

sure.